mlrun/launcher/base.py

```diff
@@ -34,17 +46,152 @@ def save(runtime):
+                        * A dictionary of configurations to use when logging. Further info per object type and artifact
+                          type can be given there. The artifact key must appear in the dictionary as "key": "the_key".
+
+        :return: run context object (RunObject) with run metadata, results and status
+        """
+        self._validate_runtime(runtime, inputs)
+        # TODO: Needed for client side
+        # self._enrich_function(runtime)
+        run = self._create_run_object(runspec)
+        return self._run(runtime, run, param_file_secrets)
+
+    @abc.abstractmethod
+    def _run(self, runtime: BaseRuntime, run: RunObject, param_file_secrets):
+        raise NotImplementedError()
+
     @staticmethod
-    def run(runtime):
-        """run the function from the server/client[local/remote]"""
-        pass
+    def _create_run_object(run: Union[str, dict, RunTemplate, RunObject]) -> RunObject:
+        # TODO: Once implemented the `Runtime` handlers configurations (doc strings, params type hints and returning
+        #       log hints, possible parameter values, etc), the configured type hints and log hints should be set into
+        #       the `RunObject` from the `Runtime`.
+        if run:
+            run = deepcopy(run)
+            if isinstance(run, str):
+                run = ast.literal_eval(run)
+            if not isinstance(run, (dict, RunTemplate, RunObject)):
+                raise ValueError(
+                    "task/runspec is not a valid task object," f" type={type(run)}"
+                )
+
+        if isinstance(run, RunTemplate):
+            run = RunObject.from_template(run)
+        if isinstance(run, dict) or run is None:
+            run = RunObject.from_dict(run)
+        return run
 
     @staticmethod
     @abc.abstractmethod
-    def _enrich_runtime(runtime):
+    def _enrich_run(runtime: BaseRuntime, run: RunObject):
         pass
 
     @staticmethod
-    @abc.abstractmethod
-    def _validate_runtime(runtime):
-        pass
+    def _store_function(runtime: BaseRuntime, run: RunObject, db):
+        run.metadata.labels["kind"] = runtime.kind
+        if "owner" not in run.metadata.labels:
+            run.metadata.labels["owner"] = (
+                environ.get("V3IO_USERNAME") or getpass.getuser()
+            )
+        if run.spec.output_path:
+            run.spec.output_path = run.spec.output_path.replace(
+                "{{run.user}}", run.metadata.labels["owner"]
+            )
+
+        if db and runtime.kind != "handler":
+            struct = runtime.to_dict()
+            hash_key = db.store_function(
+                struct, runtime.metadata.name, runtime.metadata.project, versioned=True
+            )
+            run.spec.function = runtime._function_uri(hash_key=hash_key)
+
+    @staticmethod
+    def _validate_runtime(
+        runtime: BaseRuntime,
+        inputs: Dict[str, str] = None,
```

Comment:
lets be consistent with the type hinting, if we have param that has default it means it is an optional param, lets add that type hinting to all others methods
```suggestion
        inputs: typing.Optional[Dict[str, str]] = None,
```